price_feed/lib.rs

```diff
@@ -0,0 +1,197 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[ink::contract]
+mod price_feed {
+    use ink::prelude::vec::Vec;
+
+    #[ink(storage)]
+    pub struct PriceFeed {}
+    impl PriceFeed {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+        #[ink(message)]
+        pub fn get_coin_info(
+            &mut self,
+            blockchain: [u8; 32],
+            symbol: [u8; 32],
+        ) -> Result<CoinInfo, ChainExtensionError> {
+            let input = (blockchain, symbol);
+            ::ink::env::chain_extension::ChainExtensionMethod::build(7777u32)
+                .input::<([u8; 32], [u8; 32])>()
+                .output::<Result<CoinInfo, ChainExtensionError>, false>()
+                .handle_error_code::<ChainExtensionError>()
+                .call(&input)?
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct CoinInfo {
+        pub symbol: Vec<u8>,
+        pub name: Vec<u8>,
+        pub blockchain: Vec<u8>,
+        pub supply: u128,
+        pub last_update_timestamp: u64,
+        pub price: u128,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum ChainExtensionError {
+        /// Some error occurred.
+        Other,
+        /// Failed to lookup some data.
+        CannotLookup,
+        /// A bad origin.
+        BadOrigin,
+        /// A custom error in a module.
+        Module,
+        /// At least one consumer is remaining so the account cannot be destroyed.
+        ConsumerRemaining,
+        /// There are no providers so the account cannot be created.
+        NoProviders,
+        /// There are too many consumers so the account cannot be created.
+        TooManyConsumers,
+        /// An error to do with tokens.
+        Token(PalletAssetTokenErr),
+        /// An arithmetic error.
+        Arithmetic(PalletAssetArithmeticErr),
+        // unknown error
+        Unknown,
```

Comment:
Is this `ChainExtensionError` struct copied/adopted from somewhere or are we introducing it here as we wish? How is `Other` different from `Unknown`? Why would I use one over the other?